### PR TITLE
Add details for field.Required

### PR DESCRIPTION
**Pull request by sgallagher**
Some fields are only required if other fields have certain values.
This patch adds some additional error messaging to inform the user
why these fields were necessary.

Fixes issue #7118

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

### CODE REVIEW

File: pkg/authorization/api/validation/validation.go [hunk 1]
```diff
@@ -292,7 +292,7 @@ func validateRoleBindingSubject(subject kapi.ObjectReference, isNamespaced bool,
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("name"), subject.Name, reason))
 		}
 		if !isNamespaced && len(subject.Namespace) == 0 {
-			allErrs = append(allErrs, field.Required(fldPath.Child("namespace"), ""))
+			allErrs = append(allErrs, field.Required(fldPath.Child("namespace"), "Requests without a default namespace must have one provided explicitly"))
```
Comment (deads2k) [inline pkg/authorization/api/validation/validation.go]: `ServiceAccount` subjects for `ClusterRoleBindings` must have a namespace

File: pkg/cmd/server/api/validation/etcd.go [hunk 2]
```diff
@@ -42,7 +42,7 @@ func ValidateEtcdConnectionInfo(config api.EtcdConnectionInfo, server *api.EtcdC
 		// Require a client cert to connect to an etcd that requires client certs
 		if len(server.ServingInfo.ClientCA) > 0 {
 			if len(config.ClientCert.CertFile) == 0 {
-				allErrs = append(allErrs, field.Required(fldPath.Child("certFile"), ""))
+				allErrs = append(allErrs, field.Required(fldPath.Child("certFile"), "Must provide both the certificate file and the CA file"))
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/etcd.go]: When the <path to clientCA> is set, a client certificate is required for authentication.

File: pkg/cmd/server/api/validation/validation.go [hunk 3]
```diff
@@ -78,10 +78,10 @@ func ValidateCertInfo(certInfo api.CertInfo, required bool, fldPath *field.Path)
 
 	if required || len(certInfo.CertFile) > 0 || len(certInfo.KeyFile) > 0 {
 		if len(certInfo.CertFile) == 0 {
-			allErrs = append(allErrs, field.Required(fldPath.Child("certFile"), ""))
+			allErrs = append(allErrs, field.Required(fldPath.Child("certFile"), "Both certificate and key must be provided"))
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/validation.go]: build the field path for the other element and specify it here.

Comment (deads2k) [inline pkg/cmd/server/api/validation/validation.go]: Probably want to indicate that its "both or neither"

File: pkg/image/api/validation/validation.go [hunk 4]
```diff
@@ -111,7 +111,7 @@ func ValidateImageStreamTagReference(tagRef api.TagReference, fldPath *field.Pat
 	var errs field.ErrorList
 	if tagRef.From != nil {
 		if len(tagRef.From.Name) == 0 {
-			errs = append(errs, field.Required(fldPath.Child("from", "name"), "name is required"))
+			errs = append(errs, field.Required(fldPath.Child("from", "name"), "name is required for tag reference"))
```
Comment (deads2k) [inline pkg/image/api/validation/validation.go]: Prexisting, but this looks like it adds zero entropy.  Any idea what value the string is adding?

File: pkg/user/api/validation/validation.go [hunk 5]
```diff
@@ -158,7 +158,7 @@ func ValidateIdentity(identity *api.Identity) field.ErrorList {
 		allErrs = append(allErrs, field.Invalid(userPath.Child("uid"), identity.User.UID, "may not be set if user.name is empty"))
 	}
 	if len(identity.User.Name) != 0 && len(identity.User.UID) == 0 {
-		allErrs = append(allErrs, field.Required(userPath.Child("uid"), ""))
+		allErrs = append(allErrs, field.Required(userPath.Child("uid"), "may not be set if user.UID is empty"))
```
Comment (deads2k) [inline pkg/user/api/validation/validation.go]: "Required when <field path to user.name> is set"?

File: pkg/cmd/server/api/validation/validation.go (hunk 3 above)
Comment (sgallagher) [inline pkg/cmd/server/api/validation/validation.go]: Sorry, I don't understand what you mean by "build the field path for the other element and specify it here." Are you suggesting I throw an error on both fields if either is specified without the other?

Yeah, I'll at least rephrase it to be "both or neither".

File: pkg/image/api/validation/validation.go (hunk 4 above)
Comment (sgallagher) [inline pkg/image/api/validation/validation.go]: No idea at all. Shall I remove it?

Comment (deads2k) [inline pkg/image/api/validation/validation.go]: > No idea at all. Shall I remove it?

Yes

File: pkg/authorization/api/validation/validation.go [hunk 6]
```diff
@@ -292,7 +292,7 @@ func validateRoleBindingSubject(subject kapi.ObjectReference, isNamespaced bool,
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("name"), subject.Name, reason))
 		}
 		if !isNamespaced && len(subject.Namespace) == 0 {
-			allErrs = append(allErrs, field.Required(fldPath.Child("namespace"), ""))
+			allErrs = append(allErrs, field.Required(fldPath.Child("namespace"), "ServiceAccount subjects for ClusterRoleBindings must have a namespace"))
```
Comment (smarterclayton) [inline pkg/authorization/api/validation/validation.go]: This is targeted for humans - we typically use non-Kind form ("service account") in messages intended for end users

File: pkg/cmd/server/api/validation/etcd.go [hunk 7]
```diff
@@ -42,7 +42,7 @@ func ValidateEtcdConnectionInfo(config api.EtcdConnectionInfo, server *api.EtcdC
 		// Require a client cert to connect to an etcd that requires client certs
 		if len(server.ServingInfo.ClientCA) > 0 {
 			if len(config.ClientCert.CertFile) == 0 {
-				allErrs = append(allErrs, field.Required(fldPath.Child("certFile"), ""))
+				allErrs = append(allErrs, field.Required(fldPath.Child("certFile"), "A client certificate must be provided to this etcd server"))
```
Comment (smarterclayton) [inline pkg/cmd/server/api/validation/etcd.go]: Use of "to" feels wrong here since this is a server talking to itself

File: pkg/cmd/server/api/validation/oauth.go [hunk 8]
```diff
@@ -243,13 +243,11 @@ func ValidateRequestHeaderIdentityProvider(provider *api.RequestHeaderIdentityPr
 		validationResults.AddErrors(field.Required(fieldPath.Child("provider", "headers"), ""))
 	}
 	if identityProvider.UseAsChallenger && len(provider.ChallengeURL) == 0 {
-		err := field.Required(fieldPath.Child("provider", "challengeURL"), "")
-		err.Detail = "challengeURL is required if challenge=true"
+		err := field.Required(fieldPath.Child("provider", "challengeURL"), "challengeURL is required if challenge=true")
```
Comment (smarterclayton) [inline pkg/cmd/server/api/validation/oauth.go]: Is challenge a query parameter or config value?  If the latter, say "challenge is true"

File: pkg/image/api/validation/validation.go [hunk 9]
```diff
@@ -243,7 +243,7 @@ func ValidateImageStreamImport(isi *api.ImageStreamImport) field.ErrorList {
 		switch from.Kind {
 		case "DockerImage":
 			if len(spec.From.Name) == 0 {
-				errs = append(errs, field.Required(repoPath.Child("from", "name"), ""))
+				errs = append(errs, field.Required(repoPath.Child("from", "name"), "Docker images require a name"))
```
Comment (smarterclayton) [inline pkg/image/api/validation/validation.go]: "Docker image references require a name"

**Comment by sgallagher** [comment]
@deads2k @smarterclayton Updated according to code review comments.
